compose migration (Timetable, SearchPage, Notification, LecturesOfTable)

JuTaK97 opened the pull request:
- 원활한 테스트를 위해 로컬 로그인만 임시로 연결
- ~~TimeTablePage에 Canvas size 관련 문제점 존재 (수정 필요)~~   [수정완료](https://github.com/wafflestudio/SNUTT-android/pull/65/commits/ce7d0ffbfdfd43abc9b53ed99df90634853c5faa)
- Notification, TableList는 theme 미적용 (추후 일괄 적용) #48, #55 
- HomeDrawer는 시간표 변경 테스트 용으로 최소한만 임시 적용

JuTaK97 committed:
compose migration(TimeTable, TableList, Notification, HomeDrawer)

- 원활한 테스트를 위해 로컬 로그인만 임시로 연결
- TimeTablePage에 Canvas size 관련 문제점 존재 (수정 필요)
- Notification, TableList는 theme 미적용 (추후 일괄 적용)
- HomeDrawer는 시간표 변경 테스트 용으로 최소한만 임시 적용

JuTaK97 committed:
TimeTablePage 개선

- unitWidth를 Canvas 안에서 계산하도록 변경
- SearchPage에서의 사용을 위해touchEnabled 추가
- SearchPage에서 사용하는 DrawSelectLecture 추가

JuTaK97 committed:
compose migration(SearchPage)

TODO
- EditText 구현 (+검색 필터)

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/components/compose/EditText.kt:
```diff
@@ -13,10 +17,24 @@ fun EditText(
+            if (isFocused) ExitIcon()
+            else FilterIcon()
+        },
+        keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
+        keyboardActions = KeyboardActions(
```
[nit]

EditText 자체에 Search 를 내장하는 것 보다
EditText 함수의 인자로 keybaordOptions, keyboardActions 를 노출시켜서 외부에서 핸들링하게 하는게 좋을 것 같아요

Search 가 아닌 공용적인 컴포넌트인 것 같아서요

JuTaK97 replied:
TextField 좌우에 있는 아이콘은 leadingIcon, trailingIcon 를 쓰지 말고 따로 만드는 게 좋을까요?
디자인 상 EditText의 영역(회색 사각형) 안에 들어 있는 버튼이라서 leadingIcon이랑 trailingIcon을 쓰면 모습은 깔끔하고 좋지만 터치 이벤트를 만들지 못하는 것 같아요. 그렇다고 EditText와 별개의 컴포넌트로 만들면 TextField의 각종 색이랑 옵션을 이것저것 많이 바꿔야 하는 것 같아요

JuTaK97 replied:
> Member

TextField 좌우에 있는 아이콘은 leadingIcon, trailingIcon 를 쓰지 말고 따로 만드는 게 좋을까요?
디자인 상 EditText의 영역(회색 사각형) 안에 들어 있는 버튼이라서 leadingIcon이랑 trailingIcon을 쓰면 모습은 깔끔하고 좋지만 터치 이벤트를 만들지 못하는 것 같아요. 그렇다고 EditText와 별개의 컴포넌트로 만들면 TextField의 각종 색이랑 옵션을 이것저것 많이 바꿔야 하는 것 같아요

JuTaK97 replied:
> [nit]
> 
> EditText 자체에 Search 를 내장하는 것 보다 EditText 함수의 인자로 keybaordOptions, keyboardActions 를 노출시켜서 외부에서 핸들링하게 하는게 좋을 것 같아요
> 
> Search 가 아닌 공용적인 컴포넌트인 것 같아서요

TextField 좌우에 있는 아이콘은 leadingIcon, trailingIcon 를 쓰지 말고 따로 만드는 게 좋을까요?
디자인 상 EditText의 영역(회색 사각형) 안에 들어 있는 버튼이라서 leadingIcon이랑 trailingIcon을 쓰면 모습은 깔끔하고 좋지만 터치 이벤트를 만들지 못하는 것 같아요. 그렇다고 EditText와 별개의 컴포넌트로 만들면 TextField의 각종 색이랑 옵션을 이것저것 많이 바꿔야 하는 것 같아요

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/RootActivity.kt:
```diff
@@ -49,29 +56,32 @@ class RootActivity : BaseActivity() {
 
         val startDestination =
-            if (snuttStorage.accessToken.get().isNotEmpty()) "onboard" else "home"
+            if (snuttStorage.accessToken.get().isEmpty()) "onboard" else "home"
```
[nit]
raw string 대신 이번에 만드신 NavigationDestination 사용하기?

sanggggg replied:
아래도 마찬가지로요!

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/RootActivity.kt:
```diff
@@ -85,7 +95,7 @@ class RootActivity : BaseActivity() {
             composable("signIn") {
                 SignInPage(
-                    onClickSignIn = { _, _ -> navController.navigateAsOrigin("home") },
+                    onSuccessSignIn = { navController.navigateAsOrigin("home") },
```
일시적인 코드이긴 했는데, 결국
SignInPage 에서 request & response 핸들링을 담당하게 되었으니
onSuccessSignIn 함수를 구지 여기서 내려줄 필요가 없을 것 같아요

SignInPage 내부에서 context 의 navController 를 받아와서 라우팅을 진행하는 형태로 바꾸면 어떨까요? (아래 facebook 도 마찬가지)

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomeDrawer.kt:
```diff
@@ -1,16 +1,52 @@
-fun HomeDrawer() {
-    Text(text = "drawer")
+fun HomeDrawer(
+    // TODO: CompositionalLocal 로 넘기는 형태로 바꾸기
```
CompositionLocal 을 사용하려는 이유가 무엇인가요?

개인적인 생각은 CompoistionLocal 을 사용 시 불필요한 Props Drilling 을 막아주는 효과가 function parameter 대비해서 유리하다고 생각했는데.

DTO 성격의 정보는 최대한 function parameter 를 통해 넘겨주어 순수함수화를 시키는게 preview 나 테스팅에 용이할 것 같아서 CourseBook 정도는 parameter 로 넘겨주는 지금 방식이 괜찮은 것 같아요.

(navController, DrawerController 같은 시스템에 대한 액션을 만드는 친구들만 너무 깊은 인자 전달을 막기 위해 Context 를 쓰는게 좋다고 생각 중)

JuTaK97 replied:
> CompositionLocal 을 사용하려는 이유가 무엇인가요?
> 
> 개인적인 생각은 CompoistionLocal 을 사용 시 불필요한 Props Drilling 을 막아주는 효과가 function parameter 대비해서 유리하다고 생각했는데.
> 
> DTO 성격의 정보는 최대한 function parameter 를 통해 넘겨주어 순수함수화를 시키는게 preview 나 테스팅에 용이할 것 같아서 CourseBook 정도는 parameter 로 넘겨주는 지금 방식이 괜찮은 것 같아요.
> 
> (navController, DrawerController 같은 시스템에 대한 액션을 만드는 친구들만 너무 깊은 인자 전달을 막기 위해 Context 를 쓰는게 좋다고 생각 중)

사실 별 생각 없이 파라미터는 우선 전부 compositionalLocal로 바꿔서 만들어 봤어요.
>> 트리 범위 또는 하위 계층 구조 범위로 간주되지 않는 개념에는 CompositionLocal을 사용하지 않습니다. 
>> CompositionLocal에는 적절한 기본값이 있어야 합니다


라고 문서에는 되어 있기도 하니.. 기본값이 애매하거나 하위 컴포저블이 공통으로 쓰지 않는 것들은 파라미터로 바꿔 놓겠습니다

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomeDrawer.kt:
```diff
@@ -1,16 +1,52 @@
+}
+
+@Composable
+fun TableItem(tableDto: SimpleTableDto, onClick: (String) -> Unit) {
```
private 으로 만들기?

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomeDrawer.kt:
```diff
@@ -1,16 +1,52 @@
+fun HomeDrawer(
+    // TODO: CompositionalLocal 로 넘기는 형태로 바꾸기
+    selectedCourseBook: CourseBookDto,
+    selectedCourseBookTableList: List<SimpleTableDto>,
```
```suggestion
    tableListOfSelectedCourseBook: List<SimpleTableDto>,
```

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomeDrawer.kt:
```diff
@@ -1,16 +1,52 @@
+    // TODO: CompositionalLocal 로 넘기는 형태로 바꾸기
+    selectedCourseBook: CourseBookDto,
+    selectedCourseBookTableList: List<SimpleTableDto>,
+    onClickItem: (String) -> Unit
```
```suggestion
    onClickTableItem: (String) -> Unit
```

course book click 과 헷갈림을 방지하기 위해

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomePage.kt:
```diff
@@ -23,84 +46,196 @@ enum class HomeItem(@DrawableRes val icon: Int) {
 }
-
-val HomeNavControllerContext = compositionLocalOf<NavController> {
+val TableContext = compositionLocalOf<TableContextBundle> {
+    throw RuntimeException("")
+}
+val UncheckedNotificationContext = compositionLocalOf<Boolean> {
+    throw RuntimeException("")
+}
+val SelectedLectureContext = compositionLocalOf<Optional<LectureDto>> {
+    throw RuntimeException("")
+}
+val SearchResultContext =
+    compositionLocalOf<LazyPagingItems<DataWithState<LectureDto, LectureState>>> {
+        throw RuntimeException("")
+    }
+val SearchLazyListContext = compositionLocalOf<LazyListState> {
```
상단에 적어둔 comment 를 참고해서 실제로 CompositionLocal 을 통한 인자 전달이 유리한 상황일지, 아니면 function parameter 를 통한 전달이 더 유리할지 판단이 필요할 것 같습니다

(실제로 이정도 DTO 성격의 데이터 / 전달 depth 가 적은 친구들은 그대로 함수 파라미터로 써도 된다고 생각)

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomePage.kt:
```diff
@@ -23,84 +46,196 @@ enum class HomeItem(@DrawableRes val icon: Int) {
+
+    val keyBoardController = LocalSoftwareKeyboardController.current
+
+    rxSingle { homeViewModel.getUncheckedNotificationsExist() }
```
이렇게 작성한 코드는 매 recompose 마다 때려지는데 이게 의도한 동작이 맞나요?

그렇지 않다면, (e.g. 최초 compose 시기에만 부르고 싶다면) [LaunchedEffect ](https://developer.android.com/jetpack/compose/side-effects#launchedeffect) 내부에서 해당 rxSingle 을 호출하면 좋을 것 같습니다

(애초에 저 안에서 호출한다면 rxSingle 로 전환하지 않고 그대로 try catch 하는 suspend 스타일로 작성해도 괜찮을 것 같아요)

JuTaK97 replied:
> 이렇게 작성한 코드는 매 recompose 마다 때려지는데 이게 의도한 동작이 맞나요?
> 
> 그렇지 않다면, (e.g. 최초 compose 시기에만 부르고 싶다면) [LaunchedEffect ](https://developer.android.com/jetpack/compose/side-effects#launchedeffect) 내부에서 해당 rxSingle 을 호출하면 좋을 것 같습니다
> 
> (애초에 저 안에서 호출한다면 rxSingle 로 전환하지 않고 그대로 try catch 하는 suspend 스타일로 작성해도 괜찮을 것 같아요)

기존에는 TimeTableFragment의 onViewCreated마다 부르게 되어 있어서 compose 될 때마다 부르는 게 비슷하지 않을까 생각했습니다. 그런데 확실히 지금처럼 하면 사용자의 모든 행동마다 한 번씩 불려서 과한 면이 있네요....

앱 사용 중에도 알림이 새로 왔으면 최대한 바로 알려 주는 게 좋을 것 같아서 우선 TimeTablePage에 올 때마다 부르도록 해 봤습니다

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomePage.kt:
```diff
@@ -23,84 +46,196 @@ enum class HomeItem(@DrawableRes val icon: Int) {
+                            SearchLazyListContext provides listState,
+                            SelectedLectureContext provides selectedLecture
+                        ) {
+                            SearchPage(
```
HomePage 가 너무 비대해지는 느낌이라 SearchPage 에 콜백으로 넘겨준 대부분의 함수들은 SearchPage 내부에서 처리하는게 바람직 한 것 같아요 (search page 내부에서 ViewModel 을 가져와서 실행하는 형태)

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomePage.kt:
```diff
@@ -23,84 +46,196 @@ enum class HomeItem(@DrawableRes val icon: Int) {
+                                    keyBoardController?.hide()
+                                },
+                                addLecture = { lecture, is_forced ->
+                                    selectedTimetableViewModel.addLecture(lecture, is_forced)
```
이런 rx 들은 dispose 처리가 제대로 되지 않는 것 같은데 신경써서 챙겨야 할 것 같습니다, 그리고 이런 api 의 onError 처리에 대해서도 좀 더 고민이 필요한 것 같아서
TODO 남겨주시면 제가 고민해보겠습니다

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/search/SearchFragment.kt:
```diff
@@ -86,24 +85,24 @@ class SearchFragment : BaseFragment() {
-                            hideSoftKeyboard()
-                        }
-                    )
+//                selectedTimetableViewModel.toggleLecture(it, false)
```
[nitnitnit]

어차피 실행이 닫지 않는 이전 코드들이라 수정을 하든 말든 상관이 없지만

나중에 migration 된 것을 다시 main 에 합칠 때 main 위의 작업 변경사항이랑 conflic이 생기지 않도록 리팩 이전 코드를 최대한 이전 상태로 유지시키면 어떨까요? +헷갈려서

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/search/SearchPage.kt:
```diff
@@ -1,16 +1,244 @@
+                    .background(Color(0x80000000)) // TODO: 임시
+                    .fillMaxSize()
+            ) {
+                if (loadState.refresh is LoadState.NotLoading && loadState.append.endOfPaginationReached.not() && searchResultPagingItems.itemCount < 1) {
```
[nit]
if else else 대신 when?

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/search/SearchPage.kt:
```diff
@@ -1,16 +1,244 @@
+    val contained = lectureDataWithState.state.contained
+
+    val lectureTitle = lectureDataWithState.item.course_title
+    val instructorCreditText =
```
stringResource 로 빼내기?

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/search/SearchPage.kt:
```diff
@@ -1,16 +1,244 @@
+                }
+            }
+            // TODO: 리스트 마지막 아이템은 expand & shrink 애니메이션이 아니라 그냥 생기고 사라지는 걸루.
+            // TODO: 애니메이션 속도 좀 더 빠르게
```
👍

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/search/SearchPage.kt:
```diff
@@ -1,16 +1,244 @@
+                        .fillMaxWidth(),
+                    horizontalArrangement = Arrangement.SpaceBetween
+                ) {
+                    Text(text = "강의계획서", modifier = Modifier.clicks { })
+                    Text(text = "강의평")
+                    Text(
+                        text = if (contained) "삭제하기" else "추가하기",
```
stringResources 로 추출하기

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/timetable/SelectedTimetableViewModel.kt:
```diff
@@ -31,16 +32,19 @@ class SelectedTimetableViewModel @Inject constructor(
                 val target = it.lectureList.findLast { lec -> lec.isLectureNumberEquals(lecture) }
                 if (target != null) myLectureRepository.removeLecture(lectureId = target.id)
-                else myLectureRepository.addLecture(lectureId = lecture.id, is_force)
+                else Single.never() // FIXME
```
never 이면 그대로 이 stream 이 끝나지 않게 되는거라 error 로 적당한 IllegalStateException 을 내려주는 게 좋을 것 같아요

JuTaK97 replied:
> never 이면 그대로 이 stream 이 끝나지 않게 되는거라 error 로 적당한 IllegalStateException 을 내려주는 게 좋을 것 같아요

이부분은 조금 더 말씀드릴 게 있는데, 원래 toggleLecture 였던 거를 add랑 remove로 분리해 봤습니다.
그런데 조금 이상한게 add할 때는 currentTable을 참조할 필요 없이 그냥 lecture.id 가지고 `myLectureRepository.addLecture(lectureId = lecture.id, is_force)` 하면 잘 작동 되는데 remove할 때는 안 되더라구요. (서버에서 200도 오는데 ...) 왜 currentTable에서 `.findLast`로 검색을 해서 `myLectureRepository.removeLecture()`을 해야 동작하는지 잘 이해가 안됩니다.

아무튼 위의 이유로 remove할 때는 `findLast`의 결과물(target)의 id를 이용해야 합니다. 그런데 이게 null일 일은 없는데 (toggle 시절에는 null이면 addLecture의 대상으로 판단), null인 경우의 처리는 뭘로 해야 하나요? Single. ~~ 중에 적절한 거를 뭐로 해야 할지 감이 안와서 일단 아무거나 해 놨어요

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/timetable/TimetablePage.kt:
```diff
@@ -1,34 +1,116 @@
+}
 
 @Composable
 fun TimetablePage() {
```
해당 파일은 ui 가 잘 보이니 믿음으로 가도록 하겠습니다 ㅋㅋㅋㅋㅋ 😢  (볼게 너무 많아서...)

혹시 걱정되는 부분이 있다면 알려주시면 좀 더 보도록 할게요

JuTaK97 replied:
> 해당 파일은 ui 가 잘 보이니 믿음으로 가도록 하겠습니다 ㅋㅋㅋㅋㅋ 😢 (볼게 너무 많아서...)
> 
> 혹시 걱정되는 부분이 있다면 알려주시면 좀 더 보도록 할게요

HomePage에서 selectedTimetableViewModel로부터 table이랑 theme 받아오고 subscribeAsState를 거친 State<>를 TablePage()에서 가져다 쓰게 되는데, 이때 initial 값 때문에 문제가 있습니다. 적절한 initial value가 없으면 subscribeAsState와는 다른 방법을 써야 하는지...
지금은 앱을 처음 켜면 0.5초 정도 깜빡 하고 초기값이 떴다가 로딩된 시간표로 바뀌고 있어요
![깜빡](https://user-images.githubusercontent.com/88367636/182015085-f2737980-c618-4b28-8696-6de5065d8343.gif)

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationPage.kt:
```diff
@@ -1,15 +1,152 @@
+            painter = painterResource(R.drawable.ic_warning),
+            contentDescription = "Notification Error",
+            modifier = Modifier.size(40.dp, 40.dp),
+//            colorFilter = ColorFilter.tint(Color(99, 99, 99))
```
주석을 달아놓은 이유가 무엇인가요?
지금 color 를 챙기기 애매해서라면 TODO 달아놔서 리팩 마무리할때 놓치지 않고 챙겨질 수 있도록 해 주시면 좋을 것 같아요

JuTaK97 replied:
> 주석을 달아놓은 이유가 무엇인가요? 지금 color 를 챙기기 애매해서라면 TODO 달아놔서 리팩 마무리할때 놓치지 않고 챙겨질 수 있도록 해 주시면 좋을 것 같아요

TODO 달아 놓는 거를 깜빡했네요... 꼼꼼하게 color 필요한 부분 챙겨놓겠습니다

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/table_lectures/LecturesOfTablePage.kt:
```diff
@@ -1,15 +1,187 @@
+    onClickLecture: (lecture: LectureDto) -> Unit
+) {
+    CompositionLocalProvider(
+        LocalOverScrollConfiguration provides null
```
이건 Overscroll 효과를 지우려고 넣으신걸까요?

사실 앱 내 모든 UI 에 오버스크롤 효과를 지워도 괜찮을 것 같은데 우선 고민이 필요한 것 같아서 여기만 따로 처리하기 보다 우선 지워놓고 한번에 디자이너랑 합의 후 지우는게 좋을 것 같아요 (이미 합의 완료되었다면 전체 UI 에 대해 일관성 있게 지우는 것도 고민 필요)

sanggggg reviewed (changes requested):
temp.kt 파일 지워주세요!

전반적으로 크게 이야기 할 것은

- CompositionLocal 을 언제 써야할지 좀 더 고민해보기
- string resource 도 다양한 언어 지원을 위해 지금 작업하면서 같이 strings.xml 로 빼 놓기

정도 챙겨주시면 좋을 것 같습니다

JuTaK97 committed:
EditText 범용성

JuTaK97 committed:
SignInPage, SignUpPage 개선

AuthViewModel를 직접 가지도록 변경

JuTaK97 committed:
CompositionLocal 개선

- 전달 depth가 깊고 파라미터로 전달할 것이 많은 table, trimParam, theme 외에는 모두 파라미터로 변경
- UncheckedNotification도 Rx 사용하지 않는 형태로 변경

JuTaK97 committed:
리뷰 반영

- NavigationDestination object 사용
- 변수 이름 명확히 변경
- onError 주석 추가
- Rx dispose 관리해야 하는 곳 주석 추가
- 로그인, 회원가입에서 쓸 ProgressDialog 추가

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/NavigationDestination.kt:
```diff
@@ -0,0 +1,18 @@
+package com.wafflestudio.snutt2.views
+
+object NavigationDestination {
+    const val onboard = "onboard"
+    const val tutorial = "tutorial"
+    const val home = "home"
+    const val signIn = "signIn"
+    const val signUp = "signUp"
+    const val lecturesOfTable = "lecturesOfTable"
+    const val lectureDetail = "lectures/{lecture_id}"
+    const val lectureColorSelector = "lectures/{lecture_id}/colorSelector"
+    const val notification = "notification"
+    const val appReport = "appReport"
+    const val serviceInfo = "serviceInfo"
+    const val teamInfo = "teamInfo"
+    const val timetableConfig = "timetableConfig"
+    const val userConfig = "userConfig"
```
[기록용]
맨 앞글자 대문자 사용 & 임포트를 NavigationDestination 으로 하기 
(각 필드별로 말고)

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/HomePage.kt:
```diff
@@ -23,84 +37,158 @@ enum class HomeItem(@DrawableRes val icon: Int) {
+    val searchViewModel = hiltViewModel<SearchViewModel>()
+
+    // TODO: 새 알림 체크를 어느 주기로 해야 할지? (일단 TimeTablePage에 올 때마다 확인하도록 설정)
+    LaunchedEffect(currentScreen == HomeItem.Timetable) {
```
주석이랑 싱크 맞추기?
(현재는 나갈때, 들어올 때 모두 페치 중)

JuTaK97 committed:
리뷰 반영

JuTaK97 committed:
check style
